src/c++/perf_analyzer/concurrency_worker.h

```diff
@@ -0,0 +1,119 @@
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <memory>
+#include <queue>
+#include <random>
+
+#include "load_worker.h"
+
+namespace triton { namespace perfanalyzer {
+
+// TODO REFACTOR combine IConcurrencyWorker and IRequestRateWorker interfaces
+// (must first combine threadconfigs)
+
+/// Interface for ConcurrencyWorker
+///
+class IConcurrencyWorker {
+ public:
+  struct ThreadConfig {
+    ThreadConfig(size_t thread_id)
+        : thread_id_(thread_id), concurrency_(0),
+          non_sequence_data_step_id_(thread_id), is_paused_(false)
+    {
+    }
+
+    // ID of corresponding worker thread
+    size_t thread_id_;
+    // The concurrency level that the worker should produce
+    size_t concurrency_;
+    // The current data step id in case of non-sequence model
+    size_t non_sequence_data_step_id_;
+    // Whether or not the thread is issuing new inference requests
+    bool is_paused_;
+  };
+
+  virtual void Infer(
+      std::shared_ptr<ThreadStat> thread_stat,
+      std::shared_ptr<ThreadConfig> thread_config) = 0;
+};
+
+
+/// Worker thread for the ConcurrencyManager
+///
+/// The worker maintains concurrency in different ways:
+///   For sequence models, multiple contexts must be created for multiple
+///   concurrent sequences.
+///
+///   For non-sequence models, one context can send out multiple requests
+///   at the same time. Thus it uses one single context as every infer context
+///   creates a worker thread implicitly.
+///
+class ConcurrencyWorker : public LoadWorker, public IConcurrencyWorker {
+ public:
+  ~ConcurrencyWorker() = default;
```

Comment:
There's no need to write out the declaration if it's just being assigned to default. This will happen automatically.

Same discussion from this conversation: https://github.com/triton-inference-server/client/pull/182#discussion_r1005020691

Reply:
I think this is actually better. It is more explicit. 
The opportunity to introduce bugs, especially with C++, happens when we have these assumptions in the codebase.

Reply:
I prefer they are written out. Also, they should have been declared virtual.

Reply:
There are no subclasses of ConcurrencyWorker currently so there's no need for its destructor to be declared virtual. That's only recommended for the parent classes.

And it is actually suggested to _not_ define implicitly defined operations:

https://en.cppreference.com/w/cpp/language/rule_of_three#Rule_of_zero

> If you can avoid defining default operations, do.